### ISSUE TITLE
Why is inference of model trained with sparse data slower than model trained with dense data?

### ISSUE THREAD

**Issue by Philipp3211** [dmlc/xgboost#9585]
I have the problem that my data uses lots of memory during training. So I started using sparse data to reduce footprint. However now I noticed that the model inference (with a single data point) is significantly slower (7 seconds vs 1 second). I also noticed that the resulting model is bigger on disk (11 MB vs 3 MB). What is the reason for slower inference and can I do something about it?

I already tried putting dense data into the model that was trained with sparse data, but that leads to even longer inference time (230 seconds).

**Comment by trivialfis** [comment]
Likely something went wrong when data is changed from dense to sparse. One thing come to mind is xgboost treats 0s as valid values.

**Comment by Philipp3211** [comment]
When 0s are valid values, does that mean that my sparse data is missing information? I have plenty of binary features that are mostly zeros...

I will check if maybe data is changed during sparse transformation, thank you.

**Comment by Philipp3211** [comment]
I just checked the dimensions of the data:
sparse: 1 x 4181
dense: 1 x 4137

So there actually is a discrepancy. Not sure if these few additional features can lead to a 7-fold increase in inference time though. I will have to fix this anyway, because there should be no discrepancy in dimensions whatsovever.

**Comment by trivialfis** [comment]
> does that mean that my sparse data has invalid information

Likely, your sparse matrix might removed all the 0s

**Comment by Philipp3211** [comment]
I just found out a bit more. So the features do not seem to be the problem. The additonal features that I found in my dense data came from our data set that changes with time, so they were expected. I retrained a model with both dense and sparse data with the same data set and now the number of features are the same, but the training time still is 7-fold increased in case of the model trained with sparse data.

After I found out about this I looked for more differences between those models and found that in the case of sparse training we have our XGBoost wrapped in a custom implementation of the multi-output wrapper of sklearn because we wanted to work with sparse targets which XGBoost does not support. Our customized wrapper works basically the same as sklearn's but before passing the targets to XGBoost it densifies the targets (this way there only is a subset of the targets in densified form in memory at the same time). Is there a possibility that your multi-output implementation is more performant than our slightly customized multi-output wrapper of sklearn? This could explain the speed difference.

**Comment by trivialfis** [comment]
> Is there a possibility that your multi-output implementation is more performant than our customized multi-output wrapper of sklearn

I don't know, I did not write your implementation.

**Comment by Philipp3211** [comment]
What we changed was in the fit method, we added the inline if part where the targets are being densified:
```
self.estimators_ = Parallel(n_jobs=self.n_jobs)(
    delayed(_fit_estimator)(
        self.estimator, X, y[:, i].todense() if sparse.issparse(y) else y[:, i], **routed_params.estimator.fit
    )
    for i in range(y.shape[1])
)
```

I don't think our change has any influence on inference performance though because during inference `predict` is being called and not `fit`. So I think we can assume same inference performance as sklearn's implementation.

**Comment by Philipp3211** [comment]
Alright I just did the test. I trained a model with sparse input and dense targets on a machine with enough memory, where I could then disable the multi-output wrapper while still training with sparse inputs. And yep, the result was blazing fast, even faster than the fully densily trained model. So we found the problem, it was not the sparse data, but the slow multi-output implementation of sklearn, yours seems to be a lot better...
